bridge_core_py/bids.py

```diff
@@ -63,11 +64,27 @@ def __repr__(self) -> str:
         return self.__str__()
 
 
-def is_legal(current_bid: TrickBid, new_bid: Union[TrickBid, SpecialBid]) -> bool:
+def is_legal(current_bid: TrickBid, bid_history: list[Union[TrickBid, SpecialBid]],
+             new_bid: Union[TrickBid, SpecialBid]) -> bool:
     if isinstance(new_bid, SpecialBid):
-        if current_bid is None and new_bid == SpecialBid.DOUBLE:
-            return False
-        return True
+        if new_bid == SpecialBid.PASS:
+            return True
+        n_bids = len(bid_history)
+        if new_bid == SpecialBid.DOUBLE and n_bids > 0:
+            # TrickBid was played by opponent's team
+            if isinstance(bid_history[-1], TrickBid) or (n_bids >= 3 and
+                    bid_history[-2:] == [SpecialBid.PASS, SpecialBid.PASS] and isinstance(bid_history[-3], TrickBid)):
+                return True
+        elif new_bid == SpecialBid.REDOUBLE and n_bids > 0:
+            for i in range(0, n_bids - 1):
+                # REDOUBLE was already used by player's team
+                if bid_history[n_bids - i - 1] == SpecialBid.REDOUBLE and i % 2 != 0:
+                    return False
+            # DOUBLE was used by opponent's team
+            if bid_history[-1] == SpecialBid.DOUBLE or (n_bids >= 3 and
+                    bid_history[-2:] == [SpecialBid.PASS, SpecialBid.PASS] and bid_history[-3] == SpecialBid.DOUBLE):
```

Comment:
Interesują nas 3 ostatnie bidy nie dalej. Nie ma co robić for po całej historii.

Jedyne sytuacje, które nas interesują to (nie możemy dawać R w tym samym teamie):
- D??R
- DR -> na pewno przechodzi

Możliwe poprawne sytuacje dla ??:
- DPPR
- DTDR -> sprowadza się do sytuacji na górze (DR)

Jedyne możliwe sytuacje to tylko dolny condition. Wcześniejsze nie mają sensu bo były sprawdzane kiedy inni gracze zagrywali bidy.

Jeśli pominąłem jakiś układ to wyślij w odpowiedzi.

Reply:
Ten for był bo mi się wydawało, że redouble można użyć raz na rozgrywkę. Usunąłem i teraz przechodzą tylko opisane przez ciebie przypadki, w tym ten z dołu